<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Format VCF file(s) by filtering out all variants
not satisfaying "--min-alleles 2 --max-alleles 2 --types snps"
and setting IDs
with "%CHROM:%POS:%REF:%ALT" (see https://samtools.github.io/bcftools/).
GWAS is performed on the formatted VCF file(s) by PLINK2 software
(https://www.cog-genomics.org/plink/2.0).'><title>Perform GWAS using PLINK2 (and BCFtools) — run_eggla_gwas • eggla</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform GWAS using PLINK2 (and BCFtools) — run_eggla_gwas"><meta property="og:description" content='Format VCF file(s) by filtering out all variants
not satisfaying "--min-alleles 2 --max-alleles 2 --types snps"
and setting IDs
with "%CHROM:%POS:%REF:%ALT" (see https://samtools.github.io/bcftools/).
GWAS is performed on the formatted VCF file(s) by PLINK2 software
(https://www.cog-genomics.org/plink/2.0).'><meta property="og:image" content="https://m.canouil.fr/eggla/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">eggla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.19.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/eggla.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/models-diagnostics.html">Models Diagnostics</a>
    <a class="dropdown-item" href="../articles/model-selection.html">Model Selection</a>
    <a class="dropdown-item" href="../articles/run-cubic-splines.html">Run The Cubic Splines (Random Cubic/Linear Splines) Analysis</a>
    <a class="dropdown-item" href="../articles/adiposity-peak-rebound.html">Adiposity Peak &amp; Adiposity Rebound Caracterisation</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mcanouil/eggla/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Perform GWAS using PLINK2 (and BCFtools)</h1>
      <small class="dont-index">Source: <a href="https://github.com/mcanouil/eggla/blob/HEAD/R/run_eggla_gwas.R" class="external-link"><code>R/run_eggla_gwas.R</code></a></small>
      <div class="d-none name"><code>run_eggla_gwas.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Format VCF file(s) by filtering out all variants
not satisfaying "--min-alleles 2 --max-alleles 2 --types snps"
and setting IDs
with "%CHROM:%POS:%REF:%ALT" (see https://samtools.github.io/bcftools/).
GWAS is performed on the formatted VCF file(s) by PLINK2 software
(https://www.cog-genomics.org/plink/2.0).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_eggla_gwas</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">results</span>,</span>
<span>  <span class="va">id_column</span>,</span>
<span>  traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"slope_.*"</span>, <span class="st">"auc_.*"</span>, <span class="st">"^AP_.*"</span>, <span class="st">"^AR_.*"</span><span class="op">)</span>,</span>
<span>  <span class="va">covariates</span>,</span>
<span>  <span class="va">vcfs</span>,</span>
<span>  <span class="va">working_directory</span>,</span>
<span>  use_info <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  bin_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>bcftools <span class="op">=</span> <span class="st">"/usr/bin/bcftools"</span>, plink2 <span class="op">=</span> <span class="st">"/usr/bin/plink2"</span><span class="op">)</span>,</span>
<span>  bcftools_view_options <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  build <span class="op">=</span> <span class="st">"38"</span>,</span>
<span>  strand <span class="op">=</span> <span class="st">"+"</span>,</span>
<span>  info_type <span class="op">=</span> <span class="st">"IMPUTE2 info score via 'bcftools +impute-info'"</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  clean <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>data</dt>
<dd><p>Path to the phenotypes stored as a CSV file.</p></dd>


<dt>results</dt>
<dd><p>Paths to the zip archives or directories generated by <code><a href="run_eggla_lmm.html">run_eggla_lmm()</a></code>
(vector of length two, one male and one female path).</p></dd>


<dt>id_column</dt>
<dd><p>Name of the column where sample/individual IDs are stored.</p></dd>


<dt>traits</dt>
<dd><p>One or multiple traits, <em>i.e.</em>, columns' names from <code>data</code>, to be analysed separately.</p></dd>


<dt>covariates</dt>
<dd><p>One or several covariates, <em>i.e.</em>, columns' names from <code>data</code>, to be used.
Binary trait should be coded as '1' and '2',
where sex must be coded: '1' = male, '2' = female, 'NA'/'0' = missing.</p></dd>


<dt>vcfs</dt>
<dd><p>Path to the "raw" VCF file(s) containing
the genotypes of the individuals to be analysed.</p></dd>


<dt>working_directory</dt>
<dd><p>Directory in which computation will occur and where output files will be saved.</p></dd>


<dt>use_info</dt>
<dd><p>A logical indicating whether to extract all informations stored in the "INFO" field.</p></dd>


<dt>bin_path</dt>
<dd><p>A named list containing the path to the PLINK2 and BCFtools binaries
For PLINK2, an URL to the binary can be provided (see https://www.cog-genomics.org/plink/2.0).</p></dd>


<dt>bcftools_view_options</dt>
<dd><p>A string or a vector of strings (which will be pass to <code><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste()</a></code>)
containing BCFtools view parameters, <em>e.g.</em>, <code>"--min-af 0.05"</code>, <code>"--exclude 'INFO/INFO &lt; 0.8'"</code>,
and/or <code>"--min-alleles 2 --max-alleles 2 --types snps"</code>.</p></dd>


<dt>build</dt>
<dd><p>Build of the genome on which the SNP is orientated. Default is "38".</p></dd>


<dt>strand</dt>
<dd><p>Orientation of the site to the human genome strand used. Should be "+" (default).</p></dd>


<dt>info_type</dt>
<dd><p>Type of information provided in the INFO column, <em>e.g.</em>, "IMPUTE2 info score via 'bcftools +impute-info'",</p></dd>


<dt>threads</dt>
<dd><p>Number of threads to be used by some BCFtools and PLINK2 commands.</p></dd>


<dt>quiet</dt>
<dd><p>A logical indicating whether to suppress the output.</p></dd>


<dt>clean</dt>
<dd><p>A logical indicating whether to clean intermediary files or not.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Path to results file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"bmigrowth"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">bmigrowth_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"bmigrowth.csv"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fwrite.html" class="external-link">fwrite</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    x <span class="op">=</span> <span class="va">bmigrowth</span>,</span></span>
<span class="r-in"><span>    file <span class="op">=</span> <span class="va">bmigrowth_csv</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">results_archives</span> <span class="op">&lt;-</span> <span class="fu"><a href="run_eggla_lmm.html">run_eggla_lmm</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    data <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"bmigrowth.csv"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      colClasses <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>character <span class="op">=</span> <span class="st">"ID"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>,</span></span>
<span class="r-in"><span>    id_variable <span class="op">=</span> <span class="st">"ID"</span>,</span></span>
<span class="r-in"><span>    age_days_variable <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>    age_years_variable <span class="op">=</span> <span class="st">"age"</span>,</span></span>
<span class="r-in"><span>    weight_kilograms_variable <span class="op">=</span> <span class="st">"weight"</span>,</span></span>
<span class="r-in"><span>    height_centimetres_variable <span class="op">=</span> <span class="st">"height"</span>,</span></span>
<span class="r-in"><span>    sex_variable <span class="op">=</span> <span class="st">"sex"</span>,</span></span>
<span class="r-in"><span>    covariates <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>    male_coded_zero <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>    random_complexity <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>    parallel <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>    parallel_n_chunks <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>    working_directory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">run_eggla_gwas</span><span class="op">(</span></span></span>
<span class="r-in"><span>    data <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"bmigrowth.csv"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      colClasses <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>character <span class="op">=</span> <span class="st">"ID"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>,</span></span>
<span class="r-in"><span>    results <span class="op">=</span> <span class="va">results_archives</span>,</span></span>
<span class="r-in"><span>    id_column <span class="op">=</span> <span class="st">"ID"</span>,</span></span>
<span class="r-in"><span>    traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"slope_.*"</span>, <span class="st">"auc_.*"</span>, <span class="st">"^AP_.*"</span>, <span class="st">"^AR_.*"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    covariates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sex"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    vcfs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"vcf"</span>, package <span class="op">=</span> <span class="st">"eggla"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      pattern <span class="op">=</span> <span class="st">"\\.vcf$|\\.vcf.gz$"</span>,</span></span>
<span class="r-in"><span>      full.names <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>,</span></span>
<span class="r-in"><span>    working_directory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    bin_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      bcftools <span class="op">=</span> <span class="st">"/usr/bin/bcftools"</span>,</span></span>
<span class="r-in"><span>      plink2 <span class="op">=</span> <span class="st">"/usr/bin/plink2"</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>,</span></span>
<span class="r-in"><span>    threads <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://mickael.canouil.fr" class="external-link">Mickaël Canouil</a>, Nicole Warrington, Kimberley Burrows, Anni Heiskala.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

